src/Network/Wai/SAML2.hs

```diff
@@ -118,14 +119,22 @@ saml2Callback cfg callback app req sendResponse = do
                 Just val -> do
                     result <- validateResponse cfg val
                     let rs = lookup "RelayState" body
+                    let r = case result of
+                              Left e -> Left e
+                              Right (assertion, inResponseTo) ->
+                                Right Result
+                                       { assertion = assertion,
+                                         relayState = rs,
+                                         inResponseTo = inResponseTo}
```

Comment:
Something like this may be nicer, to not duplicate what `>>=` already does:
```suggestion
                    let rs = lookup "RelayState" body
                    result <- validateResponse cfg val >>= \(assertion, inResponseTo) -> pure Result{ 
                        assertion = assertion,
                        relayState = rs,
                        inResponseTo = inResponseTo
                    }
```

Reply:
This doesn't quite type check, but you're right about the idea

```haskell
case x of
    Left e -> Left e
    Right r -> Right (f r)
```
Is really just `(fmap f x)`

Btw. so is `x >>= \y -> pure (f y)`

I like `(<&>)` from `Data.Functor` which is just an infix `flip fmap`, so the fragment becomes 

```haskell
                    result <- validateResponse cfg val <&>
                                  fmap (\(assertion, response) ->
                                          Result
                                          { assertion = assertion,
                                            relayState = rs,
                                            response = response})
```

Note that we need the second fmap, because `validateResponse cfg val` has type `IO (Either <...>)`